### ISSUE TITLE
visualization

### ISSUE THREAD

**Issue by qiulesun** [wuhuikai/FastFCN#5]
I'm sorry for disturbing you again. Could you give a demo to show the visualization of ground-truth or any other images in PContext such as fig. 6 in paper ?

**Comment by wuhuikai** [comment]
Please run [this script](https://github.com/wuhuikai/FastFCN/blob/d82ae165da25588933f21b368cccbcfa6b51227e/experiments/segmentation/scripts/encnet_res50_pcontext.sh#L19) to get the predicted label maps.

**Comment by qiulesun** [comment]
@wuhuikai , Thanks for your replying. How do I get segmentation ground truth of both training and val sets of PContext?

**Comment by wuhuikai** [comment]
```
import os
import numpy as np

import torch

from PIL import Image
from tqdm import trange
from detail import Detail

def get_voc_pallete(num_cls):
    n = num_cls
    pallete = [0]*(n*3)
    for j in range(0,n):
            lab = j
            pallete[j*3+0] = 0
            pallete[j*3+1] = 0
            pallete[j*3+2] = 0
            i = 0
            while (lab > 0):
                    pallete[j*3+0] |= (((lab >> 0) & 1) << (7-i))
                    pallete[j*3+1] |= (((lab >> 1) & 1) << (7-i))
                    pallete[j*3+2] |= (((lab >> 2) & 1) << (7-i))
                    i = i + 1
                    lab >>= 3
    return pallete

if not os.path.exists('gts'):
    os.makedirs('gts')

root=os.path.expanduser('~/.encoding/data')
BASE_DIR = 'VOCdevkit/VOC2010'
NUM_CLASS = 59

#from detail import mask
root = os.path.join(root, BASE_DIR)
annFile = os.path.join(root, 'trainval_merged.json')
imgDir = os.path.join(root, 'JPEGImages')
detail = Detail(annFile, imgDir, 'val')
ids = detail.getImgs()
# generate masks
_mapping = np.sort(np.array([
    0, 2, 259, 260, 415, 324, 9, 258, 144, 18, 19, 22,
    23, 397, 25, 284, 158, 159, 416, 33, 162, 420, 454, 295, 296,
    427, 44, 45, 46, 308, 59, 440, 445, 31, 232, 65, 354, 424,
    68, 326, 72, 458, 34, 207, 80, 355, 85, 347, 220, 349, 360,
    98, 187, 104, 105, 366, 189, 368, 113, 115]))
_key = np.array(range(len(_mapping))).astype('uint8')
mask_file = os.path.join(root, 'val.pth')
print('mask_file:', mask_file)

def _class_to_index(mask):
    # assert the values
    values = np.unique(mask)
    for i in range(len(values)):
        assert(values[i] in _mapping)
    index = np.digitize(mask.ravel(), _mapping, right=True)
    return _key[index].reshape(mask.shape)

def _preprocess(mask_file):
    masks = {}
    tbar = trange(len(ids))
    print("Preprocessing mask, this will take a while." + \
        "But don't worry, it only run once for each split.")
    for i in tbar:
        img_id = ids[i]
        mask = Image.fromarray(_class_to_index(detail.getMask(img_id)))
        masks[img_id['image_id']] = mask
        tbar.set_description("Preprocessing masks {}".format(img_id['image_id']))
    torch.save(masks, mask_file)
    return masks

if os.path.exists(mask_file):
    masks = torch.load(mask_file)
else:
    masks = _preprocess(mask_file)

for i in trange(len(ids)):
    img_id = ids[i]

    path = os.path.basename(img_id['file_name'])

    mask = masks[img_id['image_id']]
    mask.putpalette(get_voc_pallete(256))

    outname = os.path.splitext(path)[0] + '.png'
    mask.save(os.path.join('gts', outname))
```

**Comment by wuhuikai** [comment]
You can use above script to generate `val` labels. With minor modification, it can be used to generate `train` labels. Sorry for the bad code style :sob:

**Comment by qiulesun** [comment]
Thank you very much! The codes is clear, and I get all the grount truths of both training and val images.
